sbb_binarize/sbb_binarize.py

```diff
@@ -3,69 +3,92 @@
+        # noinspection PyUnresolvedReferences
+        cv2.imwrite(str(save_path), full_image)
+
+    def binarize_image(self, img: np.ndarray) -> np.ndarray:
+        img_last = False
+        for model, model_height, model_width, _ in self.models:
+            img_res = self.binarize_image_by_model(img, model, model_height, model_width)
+            img_last = img_last + (img_res == 0)
+        img_last = (~img_last).astype(np.uint8) * 255
+        return img_last
+
+    def binarize_image_by_model(self, img: np.ndarray, model: Any, model_height: int, model_width: int) -> np.ndarray:
         # Padded images must be multiples of model size
-        padded_image_height = math.ceil(original_image_height / self.model_height) * self.model_height
-        padded_image_width = math.ceil(original_image_width / self.model_width) * self.model_width
+        original_image_height, original_image_width, image_channels = img.shape
+
+        padded_image_height = math.ceil(original_image_height / model_height) * model_height
+        padded_image_width = math.ceil(original_image_width / model_width) * model_width
         padded_image = np.zeros((padded_image_height, padded_image_width, image_channels))
         padded_image[0:original_image_height, 0:original_image_width, :] = img[:, :, :]
 
         image_batch = np.expand_dims(padded_image, 0)  # Create the batch dimension
         patches = tf.image.extract_patches(
             images=image_batch,
-            sizes=[1, self.model_height, self.model_width, 1],
-            strides=[1, self.model_height, self.model_width, 1],
+            sizes=[1, model_height, model_width, 1],
+            strides=[1, model_height, model_width, 1],
             rates=[1, 1, 1, 1],
             padding='SAME'
         )
 
         number_of_horizontal_patches = patches.shape[1]
         number_of_vertical_patches = patches.shape[2]
         total_number_of_patches = number_of_horizontal_patches * number_of_vertical_patches
-        target_shape = (total_number_of_patches, self.model_height, self.model_width, image_channels)
+        target_shape = (total_number_of_patches, model_height, model_width, image_channels)
         # Squeeze all image patches (n, m, width, height, channels) into a single big batch (b, width, height, channels)
         image_patches = tf.reshape(patches, target_shape)
         # Normalize the image to values between 0.0 - 1.0
         image_patches = image_patches / float(255.0)
 
-        predicted_patches = self.model.predict(image_patches)
+        predicted_patches = model.predict(image_patches, verbose=0)
         # We have to manually call garbage collection and clear_session here to avoid memory leaks.
         # Taken from https://medium.com/dive-into-ml-ai/dealing-with-memory-leak-issue-in-keras-model-training-e703907a6501
-        gc.collect()
-        tf.keras.backend.clear_session()
+        #gc.collect()
+        #tf.keras.backend.clear_session()
```

Comment:
@bertsky these two lines were necessary to avoid memory leaks. Have you tested this, that without these two lines we can reliably predict thousands of images without any memory leaks?

Reply:
@apacha not thousands, no. But I did not notice an increase. Perhaps this has been fixed in recent TF versions?

Anyway, I'll make a dedicated test and report back – thanks for the comment.

Reply:
Ok, it turns out you were right! Without the explicit GC calls and session clear, I do get a leaky behaviour, and even a GPU OOM after a while. (This is with TF 2.10 / CUDA 11.7)

Do you prefer me to add a revert, or simply remove the last commit and force-push?

Reply:
I'll add another commit after merging.